source/extensions/filters/http/grpc_stats/grpc_stats_filter.cc

```diff
@@ -17,66 +19,171 @@ namespace GrpcStats {
 
 namespace {
 
+// A map from gRPC service/method name to symbolized stat names
+// for the service/method.
+//
+// The expected usage pattern is that the map is populated once,
+// and can then be queried lock-free as long as it isn't being
+// modified.
```

Comment:
This is essentially `StatNameSet`, where StatNameSet::rememberBuiltin is the init-time population, and the set can then do string lookups lock-free.

I am kind of confused by the two-level map. It looks like in the PR, RequestNames is a pair of string_veiw but we are pouplating it with two stat names. Do we have `RequestNames` as a struct defined in two different namespaces? Could we call the one that contains StatNames `RequestStatNames` if we need that?

Also I don't understand why we need to even have this map be two-level. Can't we just store the string->StatName mappings separately for services and statuses, and put them both into the StatNameSet?

Sorry if this is confusing; feel free to ping me on slack to discuss at higher bandwidth :)

Reply:
I didn't know about `StatNameSet`. I agree, this is basically what I want.

For context, I was looking at the existing code in `source/common/grpc/context_impl.cc,h` for inspiration. It looks like that code also re-invents `StatNameSet`.

Another note: `StatNameSet` uses `absl::flat_hash_map<std::string, Stats::StatName>`. It could be slightly more performant if it used `Envoy::StringMap` which allows lookup by `string_view` without implicitly constructing a `std::string`.

Reply:
Yes I confess 100% guilt on doing this pattern in a bunch of different places, then realizing later that StatNameSet is a useful abstraction.  And then not cleaning up all the places that needed it.

Reply:
I think I added `StringMap` but now I think it's not needed; just using `absl::flat_hash_map<std::string...` lets you do lookups with absl::string_view without creating a  temp std::string. I think that's built into the absl hash magic, and I think I didn't realize how magical it was at the time :)

I'm trying a simple PR to simplify StringMap now.

Reply:
..and even more important, I think, I more recently added `StatNameDynamicStorage` and `StatNameDynamicPool`, which should be used in scenarios where we really want to populate a stat-name discovered during a request. That will not take a lock, but it will consume more bytes in the stat-name, just for that token. Dynamic stat-names can be joined with symbolic ones, and you'll only pay extra bytes for the dynamic portion.

Reply:
Oh wait, what I built is not exactly the same. Here's where it gets a bit weird: From the service and method, I need a separate StatName for each so that they can be joined later, but I need to look them up by the pair. So assuming I don't do the 2-stage lookup I have now, it would be (logically) `map<std::pair<string, string>, std::pair<StatName, StatName>>`.

Reply:
@jmarantz and I discussed on slack and agreed that `StatNameSet` isn't adequate to get the service/method whitelist behavior this filter needs, so the current structure, or a similar one without the nested maps (by flattening the key) is needed.

Reply:
agreed; ping me when ready for the next round?